packages/dds/shared-object-base/src/sharedObjectBase.ts

```diff
@@ -0,0 +1,448 @@
+     */
+    constructor(
+        public id: string,
+        protected runtime: IFluidDataStoreRuntime,
+        public readonly attributes: IChannelAttributes)
+    {
+        super((eventName, error) => this.runtime.raiseContainerWarning(error));
+
+        this.handle = new SharedObjectHandle(
+            this,
+            id,
+            runtime.IFluidHandleContext);
+
+        this.logger = ChildLogger.create(
+            runtime.logger,
+            undefined,
+            { all: { sharedObjectId: uuid() } },
+        );
+
+        this._serializer = new FluidSerializer(this.runtime.channelsRoutingContext);
+
+        this.attachListeners();
+    }
+
+    private attachListeners() {
+        // Only listen to these events if not attached.
+        if (!this.isAttached()) {
+            this.runtime.once("attaching", () => {
+                // Calling this will let the dds to do any custom processing based on attached
+                // like starting generating ops.
+                this.didAttach();
+            });
+        }
+    }
+
+    /**
+     * A shared object, after construction, can either be loaded in the case that it is already part of
+     * a shared document. Or later attached if it is being newly added.
+     * @param services - Services used by the shared object
+     */
+    public async load(services: IChannelServices): Promise<void> {
+        if (this.runtime.attachState !== AttachState.Detached) {
+            this.services = services;
+        }
+        await this.loadCore(services.objectStorage);
+        if (this.runtime.attachState !== AttachState.Detached) {
+            this.attachDeltaHandler();
+        }
+    }
+
+    /**
+     * Initializes the object as a local, non-shared object. This object can become shared after
+     * it is attached to the document.
+     */
+    public initializeLocal(): void {
+        this.initializeLocalCore();
+    }
+
+    /**
+     * {@inheritDoc (ISharedObject:interface).bindToContext}
+     */
+    public bindToContext(): void {
+        if (this._isBoundToContext) {
+            return;
+        }
+
+        this._isBoundToContext = true;
+
+        // Allow derived classes to perform custom processing prior to registering this object
+        this.registerCore();
+
+        this.runtime.bindChannel(this);
+    }
+
+    /**
+     * {@inheritDoc (ISharedObject:interface).connect}
+     */
+    public connect(services: IChannelServices) {
+        this.services = services;
+        this.attachDeltaHandler();
+    }
+
+    /**
+     * {@inheritDoc (ISharedObject:interface).isAttached}
+     */
+    public isAttached(): boolean {
+        return this.services !== undefined && this.runtime.attachState !== AttachState.Detached;
+    }
+
+    /**
+     * {@inheritDoc (ISharedObject:interface).captureState}
+     */
+    public captureState(): IStateHost {
+        const state = this.captureStateCore();
+        return { summarize: async (fullTree: boolean = false) => {
```

Comment:
See my comment in SharedObject class RE usage of this._serializer.
But it feels like there is no need to wrap state returned from this.captureStateCore() - it should be exactly state returned to captureState() caller, and thus likely we can have captureState() abstract here (and no need for extra interfaces and Core function)